File: CSES Problem Set/DP/GridPath.py
mod = 10**9 + 7

grid_size = int(input())
paths = [[0] * (grid_size + 1) for _ in range(grid_size + 1)]
paths[0][0] = 1

for i in range(grid_size):
    row = input()

    for j in range(grid_size):
        if row[j] != '*':
            if i > 0:
                paths[i][j] += paths[i - 1][j]
                paths[i][j] %= mod
            if j > 0:
                paths[i][j] += paths[i][j - 1]
                paths[i][j] %= mod
        else:
            paths[i][j] = 0

print(paths[grid_size - 1][grid_size - 1] % mod if grid_size > 0 else -1)

File: Platforms/CSES Problem Set/DP/GridPath.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    const long long MOD = 1000000007;

    int n;
    cin >> n;

    vector<string> grid(n);
    for (int i = 0; i < n; i++) {
        cin >> grid[i];
    }

    vector<vector<long long>> paths(n + 1, vector<long long>(n + 1, 0));
    paths[0][0] = 1;

    for (int row = 0; row < n; row++) {
        for (int col = 0; col < n; col++) {
            if (grid[row][col] != '*') {
                if (row > 0) {
                    paths[row][col] += paths[row - 1][col];
                    paths[row][col] %= MOD;
                }
                if (col > 0) {
                    paths[row][col] += paths[row][col - 1];
                    paths[row][col] %= MOD;
                }
            } else {
                paths[row][col] = 0;
            }
        }
    }

    cout << (n > 0 ? paths[n - 1][n - 1] % MOD : -1) << endl;

    return 0;
}


